<a href="https://colab.research.google.com/github/Ratnakar2022/Predicting-sentiment-of-COVID-19-tweets/blob/main/Coronavirus_Tweet_Sentiment_Analysis_Capstone_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <b><u> Project Title : Sentiment Analysis : Predicting sentiment of COVID-19 tweets</u></b>

## <b> Problem Description </b>

### This challenge asks you to build a classification model to predict the sentiment of COVID-19 tweets.The tweets have been pulled from Twitter and manual tagging has been done then.

### The names and usernames have been given codes to avoid any privacy concerns.

### You are given the following information:
1. Location
2. Tweet At
3. Original Tweet
4. Label

# **Importing libraries**

In [ ]:
import pandas as pd
import numpy as np
import re
import string

#for preprocessing
from nltk.corpus import stopwords
from nltk.tokenize import word_tokenize
from nltk.stem import PorterStemmer
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split

#for ML
from sklearn.metrics import accuracy_score,precision_score,recall_score,classification_report,confusion_matrix,f1_score
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from nltk.stem.snowball import SnowballStemmer
from sklearn.feature_extraction.text import CountVectorizer

import warnings
warnings.filterwarnings('ignore')

In [ ]:
 # load stop words
import nltk
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))
nltk.download('punkt')
import nltk
nltk.download('averaged_perceptron_tagger')
import nltk
nltk.download('wordnet')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...


True

#Loading the Data

In [11]:
#mount the drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [12]:
#reading the data
df=pd.read_csv("/content/drive/MyDrive/AlmaBetter/Sentiment Analysis : Predicting sentiment of COVID-19 tweets/Coronavirus Tweet Sentiment Analysis-20221002T085910Z-001/Coronavirus Tweet Sentiment Analysis/Data _ Resources/Coronavirus Tweets.csv",encoding='latin-1')

In [13]:
#Exploring top 5 data
df.head()

UserName  ScreenName   Location     TweetAt  \
0      3799       48751     London  16-03-2020   
1      3800       48752         UK  16-03-2020   
2      3801       48753  Vagabonds  16-03-2020   
3      3802       48754        NaN  16-03-2020   
4      3803       48755        NaN  16-03-2020   

                                       OriginalTweet           Sentiment  
0  @MeNyrbie @Phil_Gahan @Chrisitv https://t.co/i...             Neutral  
1  advice Talk to your neighbours family to excha...            Positive  
2  Coronavirus Australia: Woolworths to give elde...            Positive  
3  My food stock is not the only one which is emp...            Positive  
4  Me, ready to go at supermarket during the #COV...  Extremely Negative

In [15]:
#Exploring bottom 5 data
df.tail()

UserName  ScreenName                      Location     TweetAt  \
41152     44951       89903  Wellington City, New Zealand  14-04-2020   
41153     44952       89904                           NaN  14-04-2020   
41154     44953       89905                           NaN  14-04-2020   
41155     44954       89906                           NaN  14-04-2020   
41156     44955       89907  i love you so much || he/him  14-04-2020   

                                           OriginalTweet           Sentiment  
41152  Airline pilots offering to stock supermarket s...             Neutral  
41153  Response to complaint not provided citing COVI...  Extremely Negative  
41154  You know it’s getting tough when @KameronWild...            Positive  
41155  Is it wrong that the smell of hand sanitizer i...             Neutral  
41156  @TartiiCat Well new/used Rift S are going for ...            Negative

In [18]:
#total number of data
df.shape

(41157, 6)

As we can see above we have total 41157 data and 6 column.

In [19]:
# getting summary of the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41157 entries, 0 to 41156
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   UserName       41157 non-null  int64 
 1   ScreenName     41157 non-null  int64 
 2   Location       32567 non-null  object
 3   TweetAt        41157 non-null  object
 4   OriginalTweet  41157 non-null  object
 5   Sentiment      41157 non-null  object
dtypes: int64(2), object(4)
memory usage: 1.9+ MB


Here we get that 4 column is Object data type and 2 column in Int data type. As we wonder how can name can be in INT data type. Well beacuse of user's privacy,the Username and ScreenName is encoded in dataset.

In [20]:
# Check for duplicated entries.
print("Duplicate entry in data:",len(df[df.duplicated()]))

Duplicate entry in data: 0


In [21]:
# Creating Function for Dtype,Unique values and Null values
def datainfo():
    temp_ps = pd.DataFrame(index=df.columns)
    temp_ps['DataType'] = df.dtypes
    temp_ps["Non-null_Values"] = df.count()
    temp_ps['Unique_Values'] = df.nunique()
    temp_ps['NaN_Values'] = df.isnull().sum()
    temp_ps['NaN_Values_Percentage'] = (temp_ps['NaN_Values']/len(df))*100 
    return temp_ps

In [23]:
datainfo()

DataType  Non-null_Values  Unique_Values  NaN_Values  \
UserName         int64            41157          41157           0   
ScreenName       int64            41157          41157           0   
Location        object            32567          12220        8590   
TweetAt         object            41157             30           0   
OriginalTweet   object            41157          41157           0   
Sentiment       object            41157              5           0   

               NaN_Values_Percentage  
UserName                    0.000000  
ScreenName                  0.000000  
Location                   20.871298  
TweetAt                     0.000000  
OriginalTweet               0.000000  
Sentiment                   0.000000

Here is the all information about dataset. Null value is not there in any column except location column.

In [26]:
# statistics info
df.describe(include='all').T

count unique  \
UserName       41157.0    NaN   
ScreenName     41157.0    NaN   
Location         32567  12220   
TweetAt          41157     30   
OriginalTweet    41157  41157   
Sentiment        41157      5   

                                                             top   freq  \
UserName                                                     NaN    NaN   
ScreenName                                                   NaN    NaN   
Location                                                  London    540   
TweetAt                                               20-03-2020   3448   
OriginalTweet  @MeNyrbie @Phil_Gahan @Chrisitv https://t.co/i...      1   
Sentiment                                               Positive  11422   

                  mean           std      min      25%      50%      75%  \
UserName       24377.0  11881.146851   3799.0  14088.0  24377.0  34666.0   
ScreenName     69329.0  11881.146851  48751.0  59040.0  69329.0  79618.0   
Location           NaN           NaN      NaN      NaN      NaN      NaN   
TweetAt            NaN           NaN      NaN      NaN      NaN      NaN   
OriginalTweet      NaN           NaN      NaN      NaN      NaN      NaN   
Sentiment          NaN           NaN      NaN      NaN      NaN      NaN   

                   max  
UserName       44955.0  
ScreenName     89907.0  
Location           NaN  
TweetAt            NaN  
OriginalTweet      NaN  
Sentiment          NaN

Describe works only for numeric data. here is every column have Object dat type.
Beacuse of user's privacy,the Username and ScreenName is encoded as Integer in dataset.so that min max does not work for us here.



# Exploratory Analysis and Visualization

In [27]:
# Getting unique values of 'TweetAt' column

df['TweetAt'].unique() 

array(['16-03-2020', '17-03-2020', '18-03-2020', '19-03-2020',
       '20-03-2020', '21-03-2020', '22-03-2020', '23-03-2020',
       '24-03-2020', '25-03-2020', '26-03-2020', '27-03-2020',
       '28-03-2020', '29-03-2020', '30-03-2020', '31-03-2020',
       '01-04-2020', '02-04-2020', '03-04-2020', '04-04-2020',
       '05-04-2020', '06-04-2020', '07-04-2020', '08-04-2020',
       '09-04-2020', '10-04-2020', '11-04-2020', '12-04-2020',
       '13-04-2020', '14-04-2020'], dtype=object)

In [32]:
# function to change the string format to date

from datetime import datetime

def change_to_timestamp(column):
  return datetime.strptime(column, '%d-%m-%Y')

In [33]:
# applying the above function to 'TweetAt' column

df['TweetAt'] = df['TweetAt'].apply(change_to_timestamp)


TypeError: ignored

In [34]:
# getting the month number from 'TweetAt' column

df['TweetMonth'] = df['TweetAt'].apply(lambda x : x.month)

In [35]:
df['TweetAt'][0]

Timestamp('2020-03-16 00:00:00')